monitoring/etcd/etcd.go

```diff
@@ -0,0 +1,189 @@
+	mutex  sync.RWMutex
+
+	followersLatency     *prometheus.GaugeVec
+	followersRaftFail    *prometheus.GaugeVec
+	followersRaftSuccess *prometheus.GaugeVec
+}
+
+// NewExporter returns an initialized ETCDExporter.
+func NewExporter(config *monitoring.ETCDConfig) (*Exporter, error) {
+	transport, err := config.NewHTTPTransport()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	client := &http.Client{
+		Transport: transport,
+		Timeout:   collectMetricsTimeout,
+	}
+
+	return &Exporter{
+		client: client,
+		config: *config,
+		followersLatency: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "followers_latency",
+			Help:      "Latency time (s) between ETCD leader and follower",
+		}, []string{"followerName"}),
+		followersRaftFail: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "followers_raft_fail",
+			Help:      "Counter of Raft RPC failed requests between ETCD leader and follower",
+		}, []string{"followerName"}),
+		followersRaftSuccess: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "followers_raft_success",
+			Help:      "Counter of Raft RPC successful requests between ETCD leader and follower",
+		}, []string{"followerName"}),
+	}, nil
+}
+
+// Describe implements prometheus.Collector.
+func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
+	e.followersLatency.Describe(ch)
+	e.followersRaftFail.Describe(ch)
+	e.followersRaftSuccess.Describe(ch)
+}
+
+func (e *Exporter) collect(ch chan<- prometheus.Metric) error {
+	for _, endpoint := range e.config.Endpoints {
+
+		var leaderStats LeaderStats
+		url := fmt.Sprintf("%v/v2/stats/leader", endpoint)
+
+		resp, err := e.client.Get(url)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		defer resp.Body.Close()
+
+		payload, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		err = json.Unmarshal(payload, &leaderStats)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if leaderStats.Message != "" {
+			// Endpoint is not a leader of ETCD cluster
+			continue
+		}
+
+		for id, follower := range leaderStats.Followers {
+			e.followersRaftSuccess.WithLabelValues(id).Set(float64(follower.RaftStats.Success))
+			e.followersRaftFail.WithLabelValues(id).Set(float64(follower.RaftStats.Fail))
+			e.followersLatency.WithLabelValues(id).Set(follower.Latency.Current)
+		}
+
+		e.followersLatency.Collect(ch)
+		e.followersRaftFail.Collect(ch)
+		e.followersRaftSuccess.Collect(ch)
+		return nil
+	}
+	return trace.Errorf("ETCD cluster has no leader")
+}
+
+// Collect fetches the stats from configured ETCD endpoint and delivers them
+// as Prometheus metrics. It implements prometheus.Collector.
+func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	e.mutex.Lock() // To protect metrics from concurrent collects.
+	defer e.mutex.Unlock()
+	if err := e.collect(ch); err != nil {
+		log.Errorf("Error collecting stats from ETCD: %v", err)
```

Comment:
better to write logs in lower-case with composability in mind: `error collecting stats ...`